jq255.md

```diff
@@ -0,0 +1,1412 @@
+defined coordinates (it is then truly "at infinity"). The Weierstraß
+equation can then be converted to a *short Weierstraß equation*
+`Y^2 = X^3 + A*X + B` by using the following:
+```
+A = (3*b - a^2)/3
+B = a*(2*a^2 - 9*b)/27
+X = x + a/3
+Y = y
+```
+
+Converting to Weierstraß coordinates and back is not required for
+implementing the jq255e and jq255s groups; such coordinates have
+furthermore special cases for the points `I` and `N`. These conversions
+may be useful in some specific contexts, e.g. to leverage preexisting
+generic implementations of (short) Weierstraß curves, if only for
+testing purposes.
+
+### Extended Coordinates
+
+In order to reduce the number of field element divisions (which are
+usually quite expensive in software implementations, relative to
+additions and multiplications), the following internal representation is
+recommended: a point `P = (e, u)` is represented by the quadruplet
+`(E:Z:U:T)` of field elements, with the following rules:
+
+  - `Z != 0` (in all cases)
+  - `e = E/Z`
+  - `u = U/Z`
+  - `u^2 = T/Z` (i.e. `U^2 = T*Z`)
+
+These coordinates are called *EZUT coordinates*.
+
+The points `I` and `N`, which represent the neutral element in the group,
+have coordinates `(Z:Z:0:0)` and `(-Z:Z:0:0)`, respectively, for any
+field element `Z != 0`. It does not matter, for correct computations, which
+of `I` or `N` is used to represent the neutral; similarly, any non-zero `Z`
+can be used.
+
+*Extended affine coordinates* are a special case of extended coordinates
+where `Z = 1` (in which case, `E = e`, `U = u` and `T = u^2`). Such
+coordinates are appropriate for use in fixed precomputed tables of
+points, e.g. small multiples of the conventional generator; setting `Z = 1`
+for such points reduces the table size in memory, and allows for some
+shortcuts in operations on these points.
+
+*Jacobian coordinates* are used transiently in point-doubling formulas.
+They are hereafter called *XWJ coordinates*. They relate to the `(x, y)`
+coordinates of curve points when using the Weierstraß equation `y^2 =
+x*(x^2 + a*x + b)`. For a point `(x, y)`, its XWJ coordinates are a
+triplet `(X:W:J)` such that:
+
+  - `W != 0`
+  - `x = W/(J^2)`
+  - `w = y/x = W/J`
+
+The points `I` and `N` can be represented on XWJ coordinates with the
+triplets `(W^2:W:0)` and `(0:W:0)`, respectively, for any `W != 0`. For
+all curve points other than `I` and `N`, the coordinate `J` is non-zero.
+In general, Jacobian coordinates do not lead to efficient and complete
+point addition formulas, but they offer fast and complete formulas for
+point doublings. Thus, a strategy for computing a sequence of several
+successive point doublings is to convert the source point from EZUT to
+XWJ coordinates, compute the doublings in XWJ, then convert the result
+back into EZUT coordinates.
+
+### Encoding and Decoding
+
+A group element, represented by the point `P = (E:Z:U:T)`, can be
+encoded unambiguously into a field element `v` as follows:
```

Comment:
This might go in the later section, but if someone chose the other point of the two points per group element as the argument to group_to_field(), will that be a different field element?

Reply:
I think not, that's what `if sgn(iZ*E) == 1:` is doing.

It's also spelled out below:

> The group element is really encoded as the `u` coordinate of the
representant point whose `e` coordinate is non-negative.

This is the part that makes the prime order group, both here and in Ristretto and in Decaf: encoding will encode the two representatives to the same value.